### PR TITLE
Track model changes with django-simple-history

### PR DESCRIPTION
**Pull request by stsewd**
This is the initial implementation of a design doc
from .com.

Similar to https://github.com/readthedocs/ethical-ad-server/pull/399

We need to run these commands during deploy:

- `python manage.py migrate`
- `python manage.py populate_history --auto`
  (we could add `--batchsize 500` or similar, default is 200)

Some additional things we could do:

- Delete private data from historical user models
  after the related model has been deleted (email specially).
- Keep track of the id or username of the user that made the changes
  after it have been deleted
  (https://django-simple-history.readthedocs.io/en/latest/historical_model.html#adding-additional-fields-to-historical-models).
  After a user is deleted `historical_user` is set to null,
  so we can't know who made the changes.

### CODE REVIEW

File: readthedocs/projects/models.py [hunk 1]
```diff
@@ -437,8 +437,8 @@ class Project(models.Model):
     )
 
     tags = TaggableManager(blank=True)
+    history = HistoricalRecords()
     objects = ProjectQuerySet.as_manager()
-    all_objects = models.Manager()
```
Comment (stsewd) [inline readthedocs/projects/models.py:441]: This wasn't being used

File: readthedocs/profiles/views.py [hunk 2]
```diff
@@ -78,8 +78,9 @@ def get_object(self):
         return self.request.user
 
     def form_valid(self, form):
-        self.request.user.delete()
+        user = self.get_object()
         logout(self.request)
+        user.delete()
```
Comment (ericholscher) [inline readthedocs/profiles/views.py:83]: Why was this change needed?

**Review by ericholscher — COMMENTED** [review]
This is a great start. I know for our other implementation we also added some change messages to the modeling to better understand how things were being edited (eg. by a user in the dashboard, or via API). That is probably worth integrating in the future, but I think a simple implementation like this is good for now. 

Is there anything else before you think this change is deployable, other than the tweaks you mentioned? 

> * Delete private data from historical user models
>   after the related model has been deleted (email specially).
> * Keep track of the id or username of the user that made the changes
>   after it have been deleted
>   (https://django-simple-history.readthedocs.io/en/latest/historical_model.html#adding-additional-fields-to-historical-models).
>   After a user is deleted `historical_user` is set to null,
>   so we can't know who made the changes.

I think we definitely want to attach the username or pk to the action, so have some information about who did it in the event of a deletion. We still have the diff right, just not attributable to the user?

File: readthedocs/profiles/views.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/profiles/views.py:83]: this was needed, it was given 500 if the user was deleted before the logout

**Comment by stsewd** [comment]
>  That is probably worth integrating in the future, but I think a simple implementation like this is good for now.

+1 on tracking changed done from the api and form.

> Is there anything else before you think this change is deployable, other than the tweaks you mentioned?

I'm finishing .com (subscriptions), but it's safe to merge now.

> We still have the diff right, just not attributable to the user?

yeah, the id is still there, so the prev/next/diff methods work.

File: readthedocs/core/forms.py [hunk 3]
```diff
@@ -41,6 +39,7 @@ def save(self, commit=True):
             user.first_name = first_name
             user.last_name = last_name
             user.save()
+            update_change_reason(user, 'Changed from: form')
```
Comment (stsewd) [inline readthedocs/core/forms.py:42]: This is kind of an special case, because this is a model form for UserProfile, not User, so we can't use inheritance like the other forms.

File: readthedocs/profiles/views.py [hunk 4]
```diff
@@ -75,11 +76,13 @@ class AccountDelete(PrivateViewMixin, SuccessMessageMixin, FormView):
     success_message = _('You have successfully deleted your account')
 
     def get_object(self):
-        return self.request.user
+        return User.objects.get(pk=self.request.user.pk)
```
Comment (stsewd) [inline readthedocs/profiles/views.py:79]: update_change_reason needs a proper instance, self.request.user was returning a LazyObject.

File: readthedocs/organizations/models.py [hunk 5]
```diff
@@ -215,10 +218,7 @@ def get_absolute_url(self):
         )
 
     def __str__(self):
-        return '{organization}/{team}'.format(
-            organization=self.organization.name,
-            team=self.name,
-        )
+        return self.name
```
Comment (stsewd) [inline readthedocs/organizations/models.py:221]: When an organization is deleted, the historical record from its teams are still there and an invalid organization pk is held, any time we want to access to `team_record.organization` it will error. When checking these models on the terminal, they will error (__str__ is called), so I'm just printing the name now.

File: readthedocs/core/history.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        blank=True,
+        null=True,
+        db_index=True,
+    )
+    extra_history_user_username = models.CharField(
+        _('username'),
+        max_length=150,
+        null=True,
+        db_index=True,
+    )
```
Comment (stsewd) [inline readthedocs/core/history.py:32]: We can track the ip and user agent here too, we have access to the request from the signal

File: readthedocs/core/forms.py (hunk 3 above)
Comment (ericholscher) [reply readthedocs/core/forms.py:42]: This should be a code comment - useful context 👍

File: readthedocs/core/mixins.py [hunk 7]
```diff
@@ -17,3 +18,47 @@ def get_context_data(self, **kwargs):
...
+
+class UpdateChangeReasonPostView:
+
+    """
+    Set the change_reason on the model changed through the POST method of this view.
+
+    Use this class for views that don't use a form, like ``DeleteView``.
+    """
+
+    change_reason = 'Changed from: form'
```
Comment (ericholscher) [inline readthedocs/core/mixins.py]: This doesn't seem super useful -- can we not provide more context here? I guess we'll see how this works in practice, but I feel like it would be more useful to know _what_ form or view? Not worth a ton of extra work though, if it isn't easy.

File: readthedocs/organizations/models.py (hunk 5 above)
Comment (ericholscher) [reply readthedocs/organizations/models.py:221]: This seems bad. This is changing the UX for the actual site I think, which we don't want to do. Is there not a way to limit this code in the historical model, instead of changing the functioning of the actual code?

File: readthedocs/profiles/views.py (hunk 4 above)
Comment (ericholscher) [reply readthedocs/profiles/views.py:79]: Hrm -- seems like this is causing more issues than I expected. I think we should probably add exceptions around the `update_change_reason` instead of changing the code in these ways :/

File: readthedocs/core/mixins.py [hunk 8]
```diff
@@ -17,3 +18,47 @@ def get_context_data(self, **kwargs):
...
+
+class UpdateChangeReasonForm:
+
+    """Set the change_reason on the model changed through this form."""
+
+    change_reason = 'Changed from: form'
+
+    def save(self, commit=True):
+        obj = super().save(commit=commit)
+        update_change_reason(obj, self.change_reason)
```
Comment (ericholscher) [inline readthedocs/core/mixins.py]: It seems like we want to put exception handling around all of these `update_change_reason` -- since it sounds like you hit multiple places where it raises exceptions. We don't want the site to break because of this, so we can skip it if needed.

**Review by ericholscher — COMMENTED** [review]
Sounds like this change is influencing our code more than I was hoping. I think we should be very defensive here to not 500, and reduce changes in the functionality and code of the site if possible.

File: readthedocs/profiles/views.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/profiles/views.py:79]: But then we won't have the correct change reason, I prefer having tests for these cases (we do have tests for deleting users). We should always pass a proper instance of an object to update_change_reason.

File: readthedocs/organizations/models.py (hunk 5 above)
Comment (stsewd) [reply readthedocs/organizations/models.py:221]: this change is only when printing things on the console, for the templates we access to the organization itself

https://github.com/readthedocs/readthedocs-corporate/blob/6ae87d271fb23262df46718d184469d134137b10/readthedocsinc/projects/forms.py#L103-L108

Which aren't a record model, so the object always has an organization. We can also add a try/except block, but since this only affects the shell not sure.

File: readthedocs/core/mixins.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/core/mixins.py]: It can be changed, but not sure if provides more context, we only have one form/view per model.

File: readthedocs/core/mixins.py (hunk 8 above)
Comment (stsewd) [reply readthedocs/core/mixins.py]: Our tests should catch those, if we add this class to a model/view that doesn't support historical records, that code is wrong (same if save doesn't return an object).

File: readthedocs/core/mixins.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/core/mixins.py]: I also think we could save this data as json (string in the bd), so is easy to add more metadate, like

```json
{"origin": "form", "class": "ProjectForm"}
```

Comment (stsewd) [reply readthedocs/core/mixins.py]: I have changed the message to include the class name and have a more structured form, like we have in logs


Screenshot 2021-07-22 at 18-21-26 Change history james bond (admin) Django site admin

File: readthedocs/profiles/views.py (hunk 4 above)
Comment (ericholscher) [reply readthedocs/profiles/views.py:79]: Tests are great, but we definitely need exception handling given that it's already caused issues in testing. There will definitely be more edge cases in production, and a 500 is way worse than missing `change_reason`.

File: readthedocs/api/v3/mixins.py [hunk 9]
```diff
@@ -1,12 +1,52 @@
...
+        return f'origin=api-v3 class={klass}'
+
+    def perform_create(self, serializer):
+        obj = serializer.save()
+        update_change_reason(obj, self.get_change_reason())
+        return obj
+
+    def perform_update(self, serializer):
+        obj = serializer.save()
+        update_change_reason(obj, self.get_change_reason())
```
Comment (ericholscher) [inline readthedocs/api/v3/mixins.py:42]: I'm blocking the PR on exception handling on each `update_change_reason`. From my previous experience with EthicalAds, and your testing, I'm pretty confident this will cause production issues and raise 500's for users. We should roll this out with code that protects against this, unless there's some strong argument not to. A warning in Sentry is much better than a 500 for a user and a hotfix deploy.

**Review by ericholscher — CHANGES_REQUESTED** [review]

**Comment by humitos** [comment]
>>  After a user is deleted `historical_user` is set to null,
>> so we can't know who made the changes.
> I think we definitely want to attach the username or pk to the action

@stsewd @ericholscher if we only save he `pk`, how do we know which user made the action? I think we want to know "who deleted all these particular objects and after that deleted their account", right?

File: readthedocs/core/mixins.py (hunk 7 above)
Comment (humitos) [reply readthedocs/core/mixins.py]: Maybe adding the URL's path here is useful as well.

**Comment by stsewd** [comment]
>  if we only save he pk, how do we know which user made the action? I think we want to know "who deleted all these particular objects and after that deleted their account", right?

We are saving extra information on the record models (the user pk and the username), we also have historical records for users, so we can point who made the changes after an account is deleted.

File: readthedocs/api/v3/mixins.py (hunk 9 above)
Comment (stsewd) [reply readthedocs/api/v3/mixins.py]: changed, what I don't like is that now tests won't catch these, as we are just logging now.

Comment (ericholscher) [reply readthedocs/api/v3/mixins.py]: We can write tests to catch this logic, can't we? Making the code raise more exceptions so that it's easier to test is definitely a trade off against making the code less prone to breaking in production. I don't feel comfortable making that trade, as we can certainly find ways to catch these log messages in tests can't we?

**Review by ericholscher — APPROVED** [review]
I feel better merging this with the more defensive code. I think we can iterate a bit more on this once we have real data from production 👍 

Let's make sure we have the right release notes to make sure we deploy it smoothly.

**Comment by stsewd** [comment]
@ericholscher did you change the `--batchsize` when running `populate-history --auto` on the adserver?

File: readthedocs/core/mixins.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/core/mixins.py]: could also be an extra field to populate it from the signal since we have access to the request there.

**Comment by ericholscher** [comment]
I don't believe so -- adserver DB is tiny tho.
